common/icat/query.py

```diff
@@ -74,53 +74,52 @@ def execute_query(self, client, return_json_formattable=False):
-                # attribute name within an ICAT query
-                if isinstance(value, list):
-                    for sub_value in value:
-                        self.check_attribute_name_for_distinct(key, sub_value)
-                elif isinstance(value, str):
-                    self.check_attribute_name_for_distinct(key, value)
+            distinct_attributes = self.iterate_query_conditions_for_distinctiveness()
+            mapped_distinct_fields = self.map_distinct_attributes_to_entity_names(
+                distinct_attributes, flat_query_includes
+            )
             log.debug(
-                "Attribute names used in the distinct filter, as captured by the"
-                " query's conditions %s",
-                self.attribute_names,
+                "Attribute names used in the distinct filter, mapped to the entity they"
+                " are a part of: %s",
+                mapped_distinct_fields,
             )
-        else:
-            distinct_filter_flag = False
 
         if return_json_formattable:
             log.info("Query results will be returned in a JSON format")
             data = []
 
             for result in query_result:
-                distinct_result = {}
-                dict_result = self.entity_to_dict(result, self.query.includes)
-
-                for key, value in dict_result.items():
-                    if distinct_filter_flag:
-                        # Add only the required data as per request's distinct filter
-                        # fields
-                        if key in self.attribute_names:
-                            distinct_result[key] = dict_result[key]
-
-                if distinct_filter_flag:
-                    data.append(distinct_result)
-                else:
-                    data.append(dict_result)
+                dict_result = self.entity_to_dict(
+                    result, flat_query_includes, mapped_distinct_fields
+                )
+                data.append(dict_result)
             return data
         else:
             log.info("Query results will be returned as ICAT entities")
             return query_result
 
-    def check_attribute_name_for_distinct(self, key, value):
+    def iterate_query_conditions_for_distinctiveness(self):
+        distinct_attributes = []
+        for attribute_name, where_statement in self.query.conditions.items():
+            if isinstance(where_statement, list):
+                for sub_value in where_statement:
+                    self.check_attribute_name_for_distinct(
+                        distinct_attributes, attribute_name, sub_value
+                    )
+            elif isinstance(where_statement, str):
+                self.check_attribute_name_for_distinct(
+                    distinct_attributes, attribute_name, where_statement
+                )
+
+        return distinct_attributes
+
+    def check_attribute_name_for_distinct(self, attribute_list, key, value):
```

Comment:
It's not a big issue - just a matter of style then in this case. I think including this in-line helps readability and, since the repetition is situated very closely, is less likely to suffer from the problem that motivates DRY code: that you edit one copy and miss the other. I would fall on the side of WET code in this case: Write Everything Twice (but no more).